Remove CL_TO_AF_ERROR macro

umar456 opened the pull request:
This removes the CL_TO_AF_ERROR macro and the try/catch blocks from the OpenCL backend. This is done by redefining the CATCHALL macro to look for cl::Error exceptions in addition to the standard ArrayFire Exceptions.

shehzan10 commented:
build arrayfire osx cuda ci
build arrayfire windows ci

pavanky commented on src/api/c/err_common.hpp:
```diff
@@ -207,3 +207,7 @@ void print_error(const std::string &msg);
+
+#ifdef AF_OPENCL
+#include <err_opencl.hpp>
+#endif
```
This is a circular dependency. `err_opencl.hpp` is including `err_common.hpp`. Not sure if this is a good idea. I think the CATCHALL should go in a separate file.

pavanky commented on src/backend/opencl/err_opencl.hpp:
```diff
@@ -14,26 +14,31 @@
+  catch(...) {                                              \
+    return processException();                              \
+  }
+
```
May be instead of doing `#undef CATCHALL` and defining the whole macro to include the original `CATCHALL` functionaliy, wouldn't it be better to do the following:

```
#define COMMON_CATCHAL CATCHALL
#undef CATCHALL

#define CATCHALL \
   catch (const cl::Error &err) {\
       //handle the error \
    } \
   COMMON_CATCHALL
```

This is to ensure any changes we make in `err_common.hpp` will be reflected immediately here.

umar456 replied:
Good idea. This is a better approach.

umar456 commented:
Addressed all feedback.

Moved the logic for catching cl::Error into err_common.cpp. This seems to make the most sense and does not require additional macros to handle the same behavior.

@pavanky Yeah, this would provide us with less information if the user is running into an issue regarding the exception. We can ask them to run the application using gdb and break the program on exception (using the `catch throw` command).

BTW. This change reduced the binary size by 7MB

pavanky commented:
looks good 👍

shehzan10 commented:
build arrayfire osx opencl ci
build arrayfire linux ci
build arrayfire windows ci
